File: metal/paddle-mobile/paddle-mobile/Operators/BoxcoderOp.swift
///* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License. */

import Foundation

class BoxcoderParam<P: PrecisionType>: OpParam {
  typealias ParamPrecisionType = P
  required init(opDesc: OpDesc, inScope: Scope) throws {
    do {
      priorBox = try BoxcoderParam.getFirstTensor(key: "PriorBox", map: opDesc.inputs, from: inScope)
      priorBoxVar = try BoxcoderParam.getFirstTensor(key: "PriorBoxVar", map: opDesc.inputs, from: inScope)
      targetBox = try BoxcoderParam.getFirstTensor(key: "TargetBox", map: opDesc.inputs, from: inScope)
      output = try BoxcoderParam.getFirstTensor(key: "OutputBox", map: opDesc.outputs, from: inScope)
      codeType = try BoxcoderParam.getAttr(key: "code_type", attrs: opDesc.attrs)
      boxNormalized = try BoxcoderParam.getAttr(key: "box_normalized", attrs: opDesc.attrs)
    } catch let error {
      throw error
    }
    assert(priorBox.transpose == [0, 1, 2, 3])
    assert(priorBoxVar.transpose == [0, 1, 2, 3])
    assert(targetBox.transpose == [0, 1, 2, 3])
    assert(codeType == "decode_center_size") // encode_center_size is not implemented
    assert((targetBox.tensorDim.cout() == 3) && (targetBox.tensorDim[0] == 1)) // N must be 1 (only handle batch size = 1)
  }
  let priorBox: Texture<P>
  let priorBoxVar: Texture<P>
  let targetBox: Texture<P>
  var output: Texture<P>
  let codeType: String
  let boxNormalized: Bool
}

class BoxcoderOp<P: PrecisionType>: Operator<BoxcoderKernel<P>, BoxcoderParam<P>>, Runable, Creator, InferShaperable{
  
  typealias OpType = BoxcoderOp<P>

  func inferShape() {
    //        para.output.dim = para.input.dim
  }
  
  func runImpl(device: MTLDevice, buffer: MTLCommandBuffer) throws {
    do {
      try kernel.compute(commandBuffer: buffer, param: para)
    } catch let error {
      throw error
    }
  }
  
  func delogOutput() {
    print(" \(type) output: ")
//    let priorBoxpadToFourDim = para.priorBox.padToFourDim
//    let priorBoxArray: [Float32] = para.priorBox.metalTexture.realNHWC(dim: (n: priorBoxpadToFourDim[0], h: priorBoxpadToFourDim[1], w: priorBoxpadToFourDim[2], c: priorBoxpadToFourDim[3]))
//    print(" prior box ")
//    print(priorBoxArray.strideArray())
//
//    let priorBoxVarpadToFourDim = para.priorBoxVar.padToFourDim
//    let priorBoxVarArray: [Float32] = para.priorBoxVar.metalTexture.realNHWC(dim: (n: priorBoxVarpadToFourDim[0], h: priorBoxVarpadToFourDim[1], w: priorBoxVarpadToFourDim[2], c: priorBoxVarpadToFourDim[3]))
//    print(" prior box var ")
//    print(priorBoxVarArray.strideArray())
//
//    let targetBoxpadToFourDim = para.targetBox.padToFourDim
//    let targetBoxArray: [Float32] = para.targetBox.metalTexture.realNHWC(dim: (n: targetBoxpadToFourDim[0], h: targetBoxpadToFourDim[1], w: targetBoxpadToFourDim[2], c: targetBoxpadToFourDim[3]))
//    print(" target box ")
//    print(targetBoxArray.strideArray())
    
    let targetBoxpadToFourDim = para.targetBox.padToFourDim
    let targetBoxArray = para.targetBox.metalTexture.realNHWC(dim: (n: targetBoxpadToFourDim[0], h: targetBoxpadToFourDim[1], w: targetBoxpadToFourDim[2], c: targetBoxpadToFourDim[3]))
    print(" target box ")
    print(targetBoxArray.strideArray())
    
    let padToFourDim = para.output.padToFourDim
    let outputArray: [Float32] = para.output.metalTexture.realNHWC(dim: (n: padToFourDim[0], h: padToFourDim[1], w: padToFourDim[2], c: padToFourDim[3]))
    print(" output ")
    print(outputArray.strideArray())
  }
  
}

File: lite/operators/box_coder_op.cc
// Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "lite/operators/box_coder_op.h"
#include "lite/core/op_registry.h"

namespace paddle {
namespace lite {
namespace operators {

bool BoxCoderOpLite::CheckShape() const {
  CHECK_OR_FALSE(param_.prior_box);
  CHECK_OR_FALSE(param_.target_box);
  CHECK_OR_FALSE(param_.proposals);

  auto prior_box_dims = param_.prior_box->dims();
  CHECK_OR_FALSE(prior_box_dims.size() == 2);
  CHECK_OR_FALSE(prior_box_dims[1] == 4);
  if (param_.prior_box_var != nullptr) {
    auto box_var_dim = param_.prior_box_var->dims();
    CHECK_OR_FALSE(box_var_dim.size() == 2);
    CHECK_OR_FALSE(box_var_dim == prior_box_dims);
  }
  return true;
}

bool BoxCoderOpLite::InferShapeImpl() const {
  auto prior_box_dims = param_.prior_box->dims();
  auto target_box_dims = param_.target_box->dims();
  std::string code_type = param_.code_type;
  int axis = param_.axis;
  CHECK_OR_FALSE(code_type == "encode_center_size" ||
                 code_type == "decode_center_size");

  if (code_type == "encode_center_size") {
    CHECK_OR_FALSE(target_box_dims.size() == 2);
    CHECK_OR_FALSE(target_box_dims[1] == 4);
    param_.proposals->Resize({target_box_dims[0], prior_box_dims[0], 4});
  } else if (code_type == "decode_center_size") {
    CHECK_OR_FALSE(target_box_dims.size() == 3);
    CHECK_OR_FALSE(axis == 0 || axis == 1);
    if (axis == 0) {
      CHECK_OR_FALSE(target_box_dims[1] == prior_box_dims[0]);
    } else if (axis == 1) {
      CHECK_OR_FALSE(target_box_dims[0] == prior_box_dims[0]);
    }
    CHECK_OR_FALSE(target_box_dims[2] == prior_box_dims[1]);
    param_.proposals->Resize(target_box_dims);
  }
  if (code_type == "decode_center_size" && axis == 1) {
    param_.proposals->set_lod(param_.prior_box->lod());
  } else {
    param_.proposals->set_lod(param_.target_box->lod());
  }
  return true;
}

bool BoxCoderOpLite::AttachImpl(const cpp::OpDesc& opdesc, lite::Scope* scope) {
  auto Prior_box_name = opdesc.Input("PriorBox").front();
  auto Target_box_name = opdesc.Input("TargetBox").front();
  auto Output_box_name = opdesc.Output("OutputBox").front();
  param_.prior_box = GetVar<lite::Tensor>(scope, Prior_box_name);
  param_.target_box = GetVar<lite::Tensor>(scope, Target_box_name);
  param_.proposals = GetMutableVar<lite::Tensor>(scope, Output_box_name);
  // optional params
  std::vector<std::string> input_arg_names = opdesc.InputArgumentNames();
  if (std::find(input_arg_names.begin(),
                input_arg_names.end(),
                "PriorBoxVar") != input_arg_names.end()) {
    auto box_var_arguments = opdesc.Input("PriorBoxVar");
    if (box_var_arguments.size() > 0) {
      auto* box_var = scope->FindVar(box_var_arguments.front());
      if (box_var != nullptr) {
        param_.prior_box_var = box_var->GetMutable<Tensor>();
      }
    }
  }

  param_.code_type = opdesc.GetAttr<std::string>("code_type");
  param_.box_normalized = opdesc.GetAttr<bool>("box_normalized");
  if (opdesc.HasAttr("axis")) {
    param_.axis = opdesc.GetAttr<int>("axis");
  }

  if (opdesc.HasAttr("variance")) {
    param_.variance = opdesc.GetAttr<std::vector<float>>("variance");
  }
  return true;
}

}  // namespace operators
}  // namespace lite
}  // namespace paddle

REGISTER_LITE_OP(box_coder, paddle::lite::operators::BoxCoderOpLite);
